ls1intum/Artemis
# `Programming exercises:` Validate auxiliary repositories directly in the user interface

### Checklist
#### General


#### Client

### Motivation and Context
Currently, the validation for auxiliary repositories is only done on the server-side of the application. The user is not notified in advance if improper input or settings are provided. Therefore, we add client-side validation to assist the user in configuring auxiliary repositories for programming exercises appropriately. Also, we fix the broken repository and build plan preview and the programming exercise detail view.

### Description
**Note:** When the screenshots display a yellow alert dialog, the dialog is now red.
The following issues have been addressed:
1. The tooltip for repository and build plan preview is now displayed correctly
Previously:
![aux-repo-tooltip-prev](https://user-images.githubusercontent.com/45300855/138616413-80b2b090-150f-4273-81e3-2dc276f47c1a.PNG)

Now:
![aux-repo-tooltip](https://user-images.githubusercontent.com/45300855/138616374-30025234-a29b-4582-adb6-10f83066df41.PNG)

2. Programming Exercise Detail View displays auxiliary repositories correctly
Previously:
![aux-repo-detail-prev](https://user-images.githubusercontent.com/45300855/138616431-38a18285-1383-492b-8837-de26daab1301.PNG)

Now:
![aux-repo-display](https://user-images.githubusercontent.com/45300855/138616259-2cff785c-88d6-4dca-be30-a3a0d0e2526e.PNG)

2. Added warning when two duplicate names exist in the table
![aux-repo-edit-dup-name](https://user-images.githubusercontent.com/45300855/138616278-aa865173-fead-4e43-9344-15f02c6935f8.PNG)

3. Added warning when two duplicate directories exist in the table
![aux-repo-edit-dup-direct](https://user-images.githubusercontent.com/45300855/138616284-ede4011c-971e-4ee6-a99a-e1cf0e5ad509.PNG)

4. Added warning for illegal repository names (cannot be exercise, solution, tests, auxiliary or contain special characters)
![aux-repo-illegal-name](https://user-images.githubusercontent.com/45300855/138616323-5e6469cb-81a0-4c04-afd6-6b8b243a038e.PNG)

5. Added warning for illegal checkout directories (cannot contain special characters)
![aux-repo-illegal-directory](https://user-images.githubusercontent.com/45300855/138616328-94e5f4f3-8818-404e-b998-05dd5ca60c61.PNG)

### Steps for Testing

Prerequisites:
- 1 Instructor/Editor

1. Open the programming exercise creation view
2. Insert all required information for the programming exercise
3. Click the Add Auxiliary Repository button and play around with the input fields
4. Make sure that the appropriate warnings are displayed and the generate/update button is disabled (You will be able to click it anyway, however, this is not an issue introduced by this PR)

### Review Progress





#### Code Review
#### Manual Tests
